### PR TITLE
MM-61887: Log the userID if a metric exceeds the last histogram bucket

### PR DESCRIPTION
**Pull request by agnivade**
We create a custom histogram metric that logs the userID
when the observed value is greater or equal to the last bucket value.

This allows us to start tracking the slowest users of a system
while at the same time not polluting the Prometheus metrics
by storing a userID for every observation.

https://mattermost.atlassian.net/browse/MM-61887

```release-note
NONE
```

### CODE REVIEW

File: server/channels/app/metrics.go [hunk 1]
```diff
@@ -27,60 +28,78 @@ func (a *App) RegisterPerformanceReport(rctx request.CTX, report *model.Performa
 	for _, h := range report.Histograms {
 		switch h.Metric {
 		case model.ClientTimeToFirstByte:
-			a.Metrics().ObserveClientTimeToFirstByte(commonLabels["platform"], commonLabels["agent"], h.Value/1000)
+			a.Metrics().ObserveClientTimeToFirstByte(
+				commonLabels["platform"],
+				commonLabels["agent"],
+				userID, h.Value/1000)
```
Comment (lieut-data) [inline server/channels/app/metrics.go:34]: nit: should we newline all parameters?

File: server/enterprise/metrics/histogram.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+			mlog.String("metric_name", o.metricName),
+			mlog.String("user_id", o.userID),
+			mlog.Float("observed_value", v),
+		}
+		for k, v := range o.labels {
+			fields = append(fields,
+				mlog.String("label_name", k),
+				mlog.String("label_value", v))
+		}
+		o.logger.Warn("Metric observation exceeded.", fields...)
```
Comment (lieut-data) [inline server/enterprise/metrics/histogram.go:72]: nit: do we normally log a message with a trailing period?

File: server/enterprise/metrics/histogram.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		userID:     userID,
+	}
+}
+
+func (o *WrappedObserver) Observe(v float64) {
+	if v >= o.buckets[len(o.buckets)-1] {
+		fields := []mlog.Field{
+			mlog.String("metric_name", o.metricName),
+			mlog.String("user_id", o.userID),
+			mlog.Float("observed_value", v),
```
Comment (lieut-data) [inline server/enterprise/metrics/histogram.go:65]: Worth logging the top bucket limit as well?

**Review by lieut-data — APPROVED** [review]
This is great -- if we/a customer ends up with tons of warnings logs, what's the "next step"? Is there a way to turn it off (too many logs aren't actionable)? Is there a way to tune the upper limits, or do we have to ship a new build?

**Comment by agnivade** [comment]
> if we/a customer ends up with tons of warnings logs, what's the "next step"? Is there a way to turn it off (too many logs aren't actionable)? Is there a way to tune the upper limits, or do we have to ship a new build?

Let me actually reduce the scope of this to be applied just for the page load metrics for webapp/desktop - TTFB/TTLB/dom_interactive/page_load/splash_screen_end. This should keep the logs minimal. I will keep an eye on the logs when this gets released in Community. If even this gets too much, then a dedicated config knob needs to be added, otherwise disabling `ClientMetrics` should turn this off completely.

File: server/enterprise/metrics/histogram.go (hunk 2 above)
Comment (agnivade) [reply server/enterprise/metrics/histogram.go:72]: I thought of the same thing! There is no consensus here and some logs have period while some don't. Yet another thing to add to my logging guidelines.

File: server/enterprise/metrics/histogram.go (hunk 3 above)
Comment (agnivade) [reply server/enterprise/metrics/histogram.go:65]: Sure

File: server/enterprise/metrics/histogram.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+// which is later passed down to WrappedObserver.
+type HistogramVec struct {
+	*prometheus.HistogramVec
+	buckets []float64
+	fqName  string
+	logger  mlog.LoggerIFace
+}
+
+// WrappedObserver is a wrapper of prometheus.Observer which also addtionally
+// logs userIDs when the value exceeds or equal to the highest bucket.
```
Comment (agarciamontoro) [inline server/enterprise/metrics/histogram.go:22]: Nit: other typo and wording
```suggestion
// WrappedObserver is a wrapper of prometheus.Observer which additionally
// logs userIDs when the value exceeds or equals the one in the highest bucket.
```

**Review by agarciamontoro — APPROVED** [review]
Code looks good to me, quite elegant! But I lack a bit of context: why do we need this? When can this happen?

**Comment by agnivade** [comment]
The context starts from https://github.com/mattermost/mattermost/pull/29332. We want to understand the performance of page load, and track regressions accurately. That PR was the first step.

Now the problem with client side metrics is that everything gets bucketed together into a bunch from different users. This can be problematic because we need to understand how the page load looks for a single user throughout several days, rather than aggregated data from all users. However, we cannot simply add user_id as a label in Prometheus. So this is a first step where we start logging userids for values which exceed the highest bucket value, so that we know who are the worst affected users.

The final step is: https://mattermost.atlassian.net/browse/MM-61888

**Comment by agarciamontoro** [comment]
Thanks for the detailed context, Agniva :)

**Pull request merged**